File: lib/src/memory/memory_lineage.dart
import '../crypto/kaalka_runtime.dart';

Map<String, dynamic> buildMemoryLineage(List<Map<String, dynamic>> history) {
  final lineage = <Map<String, dynamic>>[];
  String? parent;
  for (var i = 0; i < history.length; i++) {
    final item = history[i];
    final tick = item['tick'] ?? item['step'] ?? i;
    final id = computeDeterministicHash({'tick': tick, 'parent': parent});
    lineage.add({
      'id': id,
      'tick': tick,
      'parent_id': parent,
      'stable_hash': computeDeterministicHash(item),
    });
    parent = id;
  }
  return {'lineage': lineage, 'bounded': true};
}

bool verifyMemoryLineage(List<Map<String, dynamic>> lineage) {
  for (var i = 1; i < lineage.length; i++) {
    if (lineage[i]['parent_id'] != lineage[i - 1]['id']) return false;
  }
  return true;
}

File: src/memory/memoryLineage.ts
import { computeKaalkaHashPayload } from "../crypto/kaalkaRuntime.js";

export type MemoryLineageEntry = {
  id: string;
  tick: number;
  parent_id: string | null;
  stable_hash: string;
};

export function buildMemoryLineage(
  history: Array<Record<string, unknown>>,
): { lineage: MemoryLineageEntry[]; bounded: boolean } {
  const lineage: MemoryLineageEntry[] = [];
  let parent: string | null = null;
  for (const [index, item] of history.entries()) {
    const tick = Number(item.tick ?? item.step ?? index);
    const id = computeKaalkaHashPayload({ tick, kind: item.kind ?? "step", parent });
    const stable_hash = computeKaalkaHashPayload(item);
    lineage.push({ id, tick, parent_id: parent, stable_hash });
    parent = id;
  }
  return { lineage, bounded: true };
}

export function verifyMemoryLineage(lineage: MemoryLineageEntry[]): boolean {
  for (let i = 1; i < lineage.length; i++) {
    if (lineage[i]!.parent_id !== lineage[i - 1]!.id) return false;
  }
  return true;
}
